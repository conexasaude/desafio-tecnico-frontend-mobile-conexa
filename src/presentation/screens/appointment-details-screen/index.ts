export * from './appointment-details-screen';
