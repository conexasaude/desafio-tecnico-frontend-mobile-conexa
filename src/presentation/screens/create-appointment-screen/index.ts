export * from './create-appointment-screen';
