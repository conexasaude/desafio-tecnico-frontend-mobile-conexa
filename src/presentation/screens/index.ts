export * from './appointments-screen';
export * from './create-appointment-screen';
export * from './login-screen';
