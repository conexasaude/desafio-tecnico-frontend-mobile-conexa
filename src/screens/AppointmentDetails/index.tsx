import { View } from 'react-native';
import { NativeStackScreenProps } from "@react-navigation/native-stack";
import { FontAwesome } from '@expo/vector-icons';
import RootStackParamList from '../../types/rootStackParamList';
import formatsDate from '../../utils/formatsDate';
import { Container, Content, Field, FieldContainer, Header, FieldText, Label } from './styles';
import { useEffect } from 'react';

type Props = NativeStackScreenProps<RootStackParamList, 'AppointmentDetails'>;

export default function AppointmentDetails({ route, navigation }: Props) {
	const { appointment } = route.params;
	
	useEffect(() => {
		navigation.setOptions({ title: `Consulta ${appointment.id}` })
	}, [])

	return (
		<Container>
			<Header>
				<View>
					<FontAwesome name="user-circle-o" size={72} color="black" />
				</View>
				<FieldContainer>
					<Field>
						<Label>Paciente:</Label>
						<FieldText>{appointment.paciente}</FieldText>
					</Field>

					<Field>
						<Label>Data e hora: </Label>
						<FieldText>{`${formatsDate(appointment.dataConsulta, 'weekday-date-time')}`}</FieldText>
					</Field>
				</FieldContainer>
			</Header>
				
			<Content>
				<Field>
					<Label>Observação: </Label>
					<FieldText>{appointment.observacao}</FieldText>
				</Field>
			</Content>
		</Container>
    )
}