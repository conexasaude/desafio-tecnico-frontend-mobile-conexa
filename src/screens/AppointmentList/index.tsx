import { FlatList, View } from 'react-native';
import { NativeStackScreenProps } from "@react-navigation/native-stack";
import { useEffect, useState } from 'react';
import Spinner from 'react-native-loading-spinner-overlay';
import { getAppointments } from '../../services/appointment';
import { Appointment } from '../../types/appointment';
import RootStackParamList from '../../types/rootStackParamList';
import AppointmentCard from '../../components/AppointmentCard';
import SearchInput from '../../components/SearchInput';
import { Container } from './styles';

type Props = NativeStackScreenProps<RootStackParamList, 'AppointmentList'>;

export default function AppointmentList({ navigation }: Props) {
	const [appointments, setAppointments] = useState<Appointment[]>([])
	const [loading, setLoading] = useState(false)
	const [search, setSearch] = useState("")

	async function fetchAppointments() {
		setLoading(true)
		const response = await getAppointments()
		setAppointments(response)
		setLoading(false)
	}

	async function navigateToDetails(appointment: Appointment) {
		navigation.navigate('AppointmentDetails', {
			appointment
		})
	}

	function filterAppointments(appointments: Appointment[]) {
		if(search !== '') {
			return appointments.filter(appointment => 
				String(appointment.id).toLocaleLowerCase().includes(search.toLocaleLowerCase())  || appointment.paciente.toLocaleLowerCase().includes(search.toLocaleLowerCase())
			)
		} else {
			return appointments
		}
	}
    
	useEffect(() => {
		fetchAppointments()
	}, [])

	return (
		<>
			<Spinner
				visible={loading}
				textContent={'Carregando...'}
				textStyle={{color: '#FFF'}}
			/>
				
			<Container>
				<SearchInput value={search} onChangeText={setSearch} />

				<FlatList
					data={filterAppointments(appointments)}
					renderItem={({item}) => 
						<AppointmentCard
							id={String(item.id)}
							onPress={() => navigateToDetails(item)} 
							pacient={item.paciente}
							date={item.dataConsulta}
						/>
					}
					style={{ paddingHorizontal: 16 }}
				>
				</FlatList>
			</Container>
		</>
	)
}