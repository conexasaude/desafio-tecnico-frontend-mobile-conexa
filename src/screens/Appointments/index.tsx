import React, { useEffect } from 'react';
import { useReduxSelector } from '@/hooks/useReduxSelector';
import { useReduxDispatch } from '@/hooks/useReduxDispatch';
import { FlatList } from 'react-native';
import { Card } from '@/components/Card';
import { useTheme } from '@emotion/react';
import { useNavigation } from '@react-navigation/native';
import { NavPropsLogged } from '@/routes/logged';
import { fetchAppointments } from '@/store/thunks/appointment.thunk';
import { useToast } from 'react-native-toast-notifications';
import * as Styled from './styles';

export function AppointmentsScreen() {
  const { show } = useToast();
  const { effects } = useTheme();
  const { navigate } = useNavigation<NavPropsLogged>();
  const { appointments } = useReduxSelector(state => state.appointment);

  const dispatch = useReduxDispatch();

  useEffect(() => {
    dispatch(fetchAppointments({ onError: show }));
  }, [show, dispatch]);

  return (
    <Styled.SafeArea>
      <Styled.Container>
        <Styled.Column>
          <Styled.Title>Painel de Consultas</Styled.Title>
          <Styled.Line />
        </Styled.Column>

        <Styled.Row>
          <Styled.DescriptionList>{`${appointments.length} consulta(s) agendadas`}</Styled.DescriptionList>
          <Styled.ScheduleButton
            title="Agendar Consulta"
            onPress={() => navigate('ScheduleAppointment')}
          />
        </Styled.Row>

        <FlatList
          data={appointments}
          keyExtractor={({ id }) => id.toString()}
          renderItem={({ item }) => (
            <Card
              testID="appointment-card"
              title={item.paciente}
              description={item.dataConsulta}
              button={{
                title: 'Ver Detalhes',
                onPress: () => navigate('DetailsAppointment', { id: item.id }),
              }}
            />
          )}
          contentContainerStyle={{ paddingHorizontal: effects.spacing.md }}
        />
      </Styled.Container>
    </Styled.SafeArea>
  );
}
