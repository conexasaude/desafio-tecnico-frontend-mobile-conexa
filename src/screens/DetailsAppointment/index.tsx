import React from 'react';
import * as Styled from './styles';

export function DetailsAppointmentScreen() {
  return (
    <Styled.Container>
      <Styled.Title>Details Appointment Screen</Styled.Title>
    </Styled.Container>
  );
}
