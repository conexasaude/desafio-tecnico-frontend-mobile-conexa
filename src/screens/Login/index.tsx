import { useCallback, useState } from 'react';
import { Platform, TouchableWithoutFeedback, Keyboard, KeyboardAvoidingView, Alert } from 'react-native';
import { NativeStackScreenProps } from '@react-navigation/native-stack';
import { Container, Form, Title } from './styles';
import RootStackParamList from '../../types/rootStackParamList';
import Input from '../../components/Input';
import PasswordInput from '../../components/PasswordInput';
import Button from '../../components/Button';
import { useUser } from '../../hooks/useUser';
import Spinner from 'react-native-loading-spinner-overlay';
import theme from '../../theme/theme';
import Logo from '../../components/Logo';
import { useFocusEffect } from '@react-navigation/native';

type Props = NativeStackScreenProps<RootStackParamList, 'Login'>;

export default function Login({ navigation }: Props) {
  const [email, setEmail] = useState('');
  const [password, setPassword] = useState('');
  const [loading, setLoading] = useState(false);
  const { getAuthUser } = useUser();

  async function login() {
    setLoading(true)
    Keyboard.dismiss()
    try {
      await getAuthUser(email, password)

      navigation.navigate('Home')
    } catch (error) {
      Alert.alert('Erro', String(error).replace('Error:', ''))
    } finally {
      setLoading(false)
    }
  }

  useFocusEffect(
    useCallback(() => {
      return () => {
        setEmail('')
        setPassword('')
      };
    }, [])
  )

  return (
    <TouchableWithoutFeedback onPress={Keyboard.dismiss}>
      <KeyboardAvoidingView
        behavior={Platform.OS === 'ios' ? 'padding' : 'height'}
        style={{ flex: 1 }}
      >
        <Spinner
          visible={loading}
          textContent={'Carregando...'}
          textStyle={{color: theme.colors.white}}
        />
        <Container>
          <Form>
            <Logo />

            <Title>Acesse sua conta</Title>
            
            <Input
              label="Email"
              onChangeText={setEmail}
              value={email}
              placeholder='exemplo@conexa.com'
              autoCapitalize="none"
            />

            <PasswordInput
              onChangeText={setPassword}
              value={password}
              placeholder='Senha'
            />

            <Button
              testID="login-button"
              onPress={login}
              label="Login"
            />
          </Form>
        </Container>
      </KeyboardAvoidingView>
    </TouchableWithoutFeedback>
  );
}