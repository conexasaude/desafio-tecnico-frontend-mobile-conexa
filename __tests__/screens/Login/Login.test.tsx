import React from 'react';
import {
  renderWithThemeProvider,
  matchSnapshotWithProvider,
} from '@/lib/functions/Testing';

import { LoginScreen } from '@/screens/Login';
import { fireEvent, waitFor } from '@testing-library/react-native';

const mockDispatch = jest.fn();

jest.mock('@/hooks/useReduxDispatch', () => ({
  useReduxDispatch: () => mockDispatch,
}));

jest.mock('@/hooks/useReduxSelector', () => ({
  useReduxSelector: () => ({ loading: false }),
}));

describe('Login', () => {
  test('should render correctly', () => {
    renderWithThemeProvider(<LoginScreen />);
  });

  test('should matches snapshot', () => {
    matchSnapshotWithProvider(<LoginScreen />);
  });

  test('should called dispatch', () => {
    const { getByText, getByTestId } = renderWithThemeProvider(<LoginScreen />);

    const inputEmail = getByTestId('input-email');
    const inputPassword = getByTestId('input-password');
    const submitButton = getByText('Entrar');

    fireEvent.changeText(inputEmail, 'maria@gmail.com');
    fireEvent.changeText(inputPassword, '12345678');
    fireEvent.press(submitButton);

    waitFor(() => expect(mockDispatch).toHaveBeenCalled());
  });

  test('should show an error message if you don not fill in the fields', () => {
    const { getByText } = renderWithThemeProvider(<LoginScreen />);

    const submitButton = getByText('Entrar');

    fireEvent.press(submitButton);

    waitFor(() => {
      expect(getByText('O campo de e-mail é obrigatório')).toBeDefined();
      expect(getByText('O campo de senha é obrigatório')).toBeDefined();
    });
  });

  test('should show an error message if you fill in the email field with an invalid email', () => {
    const { getByText, getByTestId } = renderWithThemeProvider(<LoginScreen />);

    const inputEmail = getByTestId('input-email');
    const inputPassword = getByTestId('input-password');
    const submitButton = getByText('Entrar');

    fireEvent.changeText(inputEmail, 'invalid');
    fireEvent.changeText(inputPassword, '12345678');
    fireEvent.press(submitButton);

    waitFor(() => {
      expect(getByText('Por favor, insira um e-mail válido')).toBeDefined();
    });
  });
});
