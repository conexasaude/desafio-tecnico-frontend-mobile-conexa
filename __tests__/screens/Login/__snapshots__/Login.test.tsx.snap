// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login should matches snapshot 1`] = `
<RCTSafeAreaView
  style={
    [
      {
        "flex": 1,
      },
      undefined,
    ]
  }
>
  <View
    style={
      [
        {
          "flex": 1,
          "paddingHorizontal": 16,
          "paddingVertical": 32,
        },
        undefined,
      ]
    }
  >
    <View
      flex={1}
      style={
        [
          {
            "flex": 1,
          },
          undefined,
        ]
      }
    />
    <SvgMock
      style={
        [
          {
            "alignSelf": "center",
          },
          undefined,
        ]
      }
    />
    <View
      flex={2}
      style={
        [
          {
            "flex": 2,
          },
          undefined,
        ]
      }
    />
    <Text
      style={
        [
          {
            "color": "#2E50D4",
            "fontFamily": "Nunito-Bold",
            "fontSize": 24,
            "marginBottom": 24,
          },
          undefined,
        ]
      }
    >
      Faça Login
    </Text>
    <View
      style={
        [
          {
            "width": "100%",
          },
          undefined,
        ]
      }
    >
      <View
        style={
          [
            {},
            {},
          ]
        }
      >
        <View
          style={
            [
              {
                "borderColor": "#2E50D4",
                "borderRadius": 8,
                "borderWidth": 2,
                "height": 52,
                "marginTop": 16,
                "padding": 8,
              },
              undefined,
            ]
          }
          testID="input-container"
          textArea={false}
        >
          <Text
            style={
              [
                {
                  "backgroundColor": "#fff",
                  "color": "#2E50D4",
                  "fontFamily": "Nunito-Bold",
                  "fontSize": 14,
                  "left": 16,
                  "padding": 4,
                  "position": "absolute",
                  "top": -14,
                },
                undefined,
              ]
            }
          >
            E-mail
          </Text>
          <View
            style={
              [
                {
                  "flex": 1,
                  "flexDirection": "row",
                },
                undefined,
              ]
            }
          >
            <TextInput
              multiline={false}
              onChangeText={[Function]}
              placeholder="Ex: jose@email.com"
              placeholderTextColor="#B3B3B3"
              style={
                [
                  {
                    "color": "#333333",
                    "flex": 1,
                    "fontFamily": "Nunito-SemiBold",
                    "fontSize": 16,
                    "padding": 0,
                  },
                  undefined,
                ]
              }
              testID="input-email"
              textAlignVertical="center"
              value=""
            />
          </View>
        </View>
      </View>
      <View
        style={
          [
            {},
            {
              "marginBottom": 40,
            },
          ]
        }
      >
        <View
          style={
            [
              {
                "borderColor": "#2E50D4",
                "borderRadius": 8,
                "borderWidth": 2,
                "height": 52,
                "marginTop": 16,
                "padding": 8,
              },
              undefined,
            ]
          }
          testID="input-container"
          textArea={false}
        >
          <Text
            style={
              [
                {
                  "backgroundColor": "#fff",
                  "color": "#2E50D4",
                  "fontFamily": "Nunito-Bold",
                  "fontSize": 14,
                  "left": 16,
                  "padding": 4,
                  "position": "absolute",
                  "top": -14,
                },
                undefined,
              ]
            }
          >
            Senha
          </Text>
          <View
            style={
              [
                {
                  "flex": 1,
                  "flexDirection": "row",
                },
                undefined,
              ]
            }
          >
            <TextInput
              multiline={false}
              onChangeText={[Function]}
              placeholder="*************"
              placeholderTextColor="#B3B3B3"
              secureTextEntry={true}
              style={
                [
                  {
                    "color": "#333333",
                    "flex": 1,
                    "fontFamily": "Nunito-SemiBold",
                    "fontSize": 16,
                    "padding": 0,
                  },
                  undefined,
                ]
              }
              testID="input-password"
              textAlignVertical="center"
              value=""
            />
            <View
              accessibilityState={
                {
                  "busy": undefined,
                  "checked": undefined,
                  "disabled": undefined,
                  "expanded": undefined,
                  "selected": undefined,
                }
              }
              accessibilityValue={
                {
                  "max": undefined,
                  "min": undefined,
                  "now": undefined,
                  "text": undefined,
                }
              }
              accessible={true}
              collapsable={false}
              focusable={true}
              onClick={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                {
                  "justifyContent": "center",
                  "opacity": 1,
                }
              }
            >
              <Text
                allowFontScaling={false}
                selectable={false}
                style={
                  [
                    {
                      "color": "#2E50D4",
                      "fontSize": 24,
                    },
                    undefined,
                    {
                      "fontFamily": "Material Design Icons",
                      "fontStyle": "normal",
                      "fontWeight": "normal",
                    },
                    {},
                  ]
                }
                testID="visibility-icon"
              >
                󰈈
              </Text>
            </View>
          </View>
        </View>
      </View>
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "alignSelf": "center",
            "backgroundColor": "#2E50D4",
            "borderColor": "#2E50D4",
            "borderRadius": 8,
            "borderStyle": "solid",
            "borderWidth": 0,
            "height": 52,
            "justifyContent": "center",
            "opacity": 1,
            "padding": 8,
            "width": "100%",
          }
        }
      >
        <Text
          style={
            [
              {
                "color": "#fff",
                "fontFamily": "Nunito-Bold",
                "fontSize": 14,
                "textAlign": "center",
              },
              undefined,
            ]
          }
          variant="contained"
        >
          Entrar
        </Text>
      </View>
    </View>
    <View
      flex={3}
      style={
        [
          {
            "flex": 3,
          },
          undefined,
        ]
      }
    />
  </View>
</RCTSafeAreaView>
`;
